src/NuGetGallery/Authentication/ApiScopeEvaluator.cs

```
@@ -0,0 +1,121 @@
+        /// <param name="action">The action that the scopes being evaluated are checked for permission to do.</param>
+        /// <param name="entity">The entity that the scopes being evaluated are checked for permission to <paramref name="action"/> on.</param>
+        /// <param name="owner">
+        /// The <see cref="User"/> specified by the <see cref="Scope.OwnerKey"/> of the <see cref="Scope"/> that evaluated with <see cref="ApiScopeEvaluationResult.Success"/>.
+        /// If no <see cref="Scope"/>s evaluate to <see cref="ApiScopeEvaluationResult.Success"/>, this will be null.
+        /// </param>
+        /// <returns>A <see cref="ApiScopeEvaluationResult"/> that describes the evaluation of the .</returns>
+        /// <remarks>This method is internal because it is tested directly.</remarks>
+        internal ApiScopeEvaluationResult Evaluate<TEntity>(
+            User currentUser,
+            IEnumerable<Scope> scopes,
+            IActionRequiringEntityPermissions<TEntity> action,
+            TEntity entity,
+            Func<TEntity, string> getSubjectFromEntity,
+            params string[] requestedActions)
+        {
+            User ownerInScope = null;
+
+            if (scopes == null || !scopes.Any())
+            {
+                // Legacy V1 API key without scopes.
+                // Evaluate it as if it has an unlimited scope.
+                scopes = new[] { new Scope(ownerKey: null, subject: NuGetPackagePattern.AllInclusivePattern, allowedAction: NuGetScopes.All) };
+            }
+
+            bool success = false;
+
+            // Check that all scopes provided have the same owner scope.
+            var ownerScopes = scopes.Select(s => s.OwnerKey);
+            var firstOwnerScope = ownerScopes.FirstOrDefault();
+            if (ownerScopes.Any(o => o != firstOwnerScope))
+            {
+                throw new ArgumentException("All scopes provided must have the same owner scope.");
+            }
```

Since this would be a bug, could we just do SingleOrDefault and let it throw the InvalidOperationException? I'm assuming the exception would go to AI so we could investigate.

---

I prefer it as is because it is very clear about the error and the expectations of the code. The error is not an `InvalidOperationException`, it is an `ArgumentException` from being given invalid arguments.